File: android/app/src/main/java/com/lilru/liftr/data/PetCombatModels.kt
package com.lilru.liftr.data

import kotlin.math.pow
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class ProfileEnergyWire(
    val current: Int = 0,
    val max: Int = 5,
    @SerialName("last_refresh") val lastRefresh: String? = null,
    @SerialName("next_refresh_at") val nextRefreshAt: String? = null,
    @SerialName("regen_minutes") val regenMinutes: Int? = null
) {
    val isFull: Boolean get() = current >= max
}

object PetEnergyPricing {
    const val MAX_CAPACITY = 15

    fun upgradeCost(maxEnergy: Int): Int =
        (5000 * 2.0.pow((maxOf(maxEnergy, 5) - 5).toDouble())).toInt()
}

@Serializable
data class EnergyUpgradeResultWire(
    @SerialName("max_energy") val maxEnergy: Int,
    @SerialName("current_energy") val currentEnergy: Int,
    val cost: Int,
    @SerialName("purchase_id") val purchaseId: String
)

@Serializable
data class PetCombatRewardsWire(
    val xp: Int = 0,
    val coins: Int = 0
)

@Serializable
data class PetCombatPetSummaryWire(
    val id: String,
    @SerialName("custom_name") val customName: String? = null,
    @SerialName("pet_type") val petType: String,
    @SerialName("evolution_stage") val evolutionStage: String,
    @SerialName("current_level") val currentLevel: Int = 1,
    val rarity: String = "common",
    @SerialName("image_url") val imageUrl: String? = null,
    val name: String? = null
) {
    fun displayName(): String {
        if (!name.isNullOrBlank()) return name
        if (!customName.isNullOrBlank()) return customName
        return petType.replace('_', ' ').replaceFirstChar { it.uppercase() }
    }
}

@Serializable
data class PetCombatStatsSummaryWire(
    val health: Int = 0,
    val strength: Int = 0,
    val defense: Int = 0,
    val speed: Int = 0,
    val intelligence: Int = 0,
    val agility: Int = 0,
    val stamina: Int = 0,
    @SerialName("critical_rate") val criticalRate: Int = 0,
    val resistance: Int = 0,
    val exploration: Int = 0,
    val happiness: Int = 0
) {
    val combatStatPoolTotal: Int
        get() = health + strength + defense + speed + agility + stamina + resistance + criticalRate + intelligence + exploration
}

object PetCombatStatBalancing {
    fun isUnbalanced(attacker: PetCombatStatsSummaryWire, defender: PetCombatStatsSummaryWire): Boolean {
        val poolA = attacker.combatStatPoolTotal
        val poolB = defender.combatStatPoolTotal
        val stronger = maxOf(poolA, poolB)
        val weaker = minOf(poolA, poolB)
        return stronger * 100 > weaker * 105
    }
}

@Serializable
data class PetCombatStatBalancingWire(
    @SerialName("is_unbalanced") val isUnbalanced: Boolean = false,
    @SerialName("attacker_stat_pool") val attackerStatPool: Int = 0,
    @SerialName("defender_stat_pool") val defenderStatPool: Int = 0,
    @SerialName("attacker_pool_battle") val attackerPoolBattle: Int? = null,
    @SerialName("defender_pool_battle") val defenderPoolBattle: Int? = null,
    @SerialName("stronger_side") val strongerSide: String? = null,
    @SerialName("hardcore_buff_multiplier") val hardcoreBuffMultiplier: Double? = null,
    @SerialName("hardcore_bonus_percent") val hardcoreBonusPercent: Int? = null
)

enum class PetCombatChallengeMode(val rawValue: String) {
    BALANCED("balanced"),
    HARDCORE("hardcore");

    val disableNerfChoice: Boolean get() = this == HARDCORE

    val title: String
        get() = when (this) {
            BALANCED -> "Balanced Mode (Safe)"
            HARDCORE -> "Hardcore Mode (No Nerf)"
        }

    companion object {
        fun fromRaw(value: String?): PetCombatChallengeMode =
            entries.firstOrNull { it.rawValue == value } ?: BALANCED
    }
}

@Serializable
data class PetCombatSideWire(
    @SerialName("user_id") val userId: String,
    val username: String? = null,
    val pet: PetCombatPetSummaryWire? = null,
    val stats: PetCombatStatsSummaryWire? = null
)

@Serializable
data class PetCombatPreviewWire(
    @SerialName("can_challenge") val canChallenge: Boolean = false,
    @SerialName("block_reason") val blockReason: String? = null,
    @SerialName("cooldown_expires_at") val cooldownExpiresAt: String? = null,
    @SerialName("stat_balancing") val statBalancing: PetCombatStatBalancingWire? = null,
    val attacker: PetCombatSideWire? = null,
    val defender: PetCombatSideWire? = null,
    val energy: ProfileEnergyWire? = null
) {
    fun isStatUnbalanced(): Boolean {
        statBalancing?.isUnbalanced?.let { return it }
        val attackerStats = attacker?.stats ?: return false
        val defenderStats = defender?.stats ?: return false
        return PetCombatStatBalancing.isUnbalanced(attackerStats, defenderStats)
    }

    fun isAttackerUnderdog(): Boolean = statBalancing?.strongerSide == "defender"

    fun hardcoreBonusLabel(): String? {
        val percent = statBalancing?.hardcoreBonusPercent ?: return null
        if (percent <= 0) return null
        return "+$percent% Coins & XP"
    }

    fun blockReasonText(): String? = when (blockReason) {
        "self_challenge" -> "You cannot challenge yourself."
        "attacker_no_pet" -> "You need a hatched pet to challenge."
        "defender_no_pet" -> "This user has no pet to challenge."
        "attacker_egg" -> "Your pet must hatch before battling."
        "defender_egg" -> "This user's pet has not hatched yet."
        "no_energy" -> "No arena energy left. You regain 1 energy every 4 hours."
        "cooldown_active" -> "You recently battled this user. Try again later."
        null -> null
        else -> blockReason.replace('_', ' ').replaceFirstChar { it.uppercase() }
    }
}

@Serializable
data class PetBattlePetSnapshotWire(
    @SerialName("user_id") val userId: String,
    @SerialName("pet_instance_id") val petInstanceId: String,
    val name: String,
    @SerialName("pet_type") val petType: String,
    @SerialName("evolution_stage") val evolutionStage: String,
    val level: Int,
    val rarity: String,
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("max_hp") val maxHp: Int? = null
) {
    fun resolvedMaxHp(fallback: Int = 1): Int = maxHp?.takeIf { it > 0 } ?: maxOf(fallback, 1)
}

@Serializable
data class PetBattleTurnWire(
    val turn: Int,
    val actor: String,
    val action: String,
    val damage: Int,
    @SerialName("is_critical") val isCritical: Boolean = false,
    @SerialName("attacker_hp_after") val attackerHpAfter: Int,
    @SerialName("defender_hp_after") val defenderHpAfter: Int,
    val message: String
)

@Serializable
data class PetBattleResultSummaryWire(
    val winner: String? = null,
    @SerialName("winner_user_id") val winnerUserId: String? = null,
    @SerialName("total_turns") val totalTurns: Int = 0,
    @SerialName("is_draw") val isDraw: Boolean = false
)

@Serializable
data class PetBattleLogWire(
    val version: Int = 1,
    @SerialName("attacker_pet") val attackerPet: PetBattlePetSnapshotWire,
    @SerialName("defender_pet") val defenderPet: PetBattlePetSnapshotWire,
    val turns: List<PetBattleTurnWire> = emptyList(),
    val result: PetBattleResultSummaryWire
)

@Serializable
data class PetCombatResultWire(
    @SerialName("combat_id") val combatId: String,
    @SerialName("winner_user_id") val winnerUserId: String? = null,
    @SerialName("attacker_rewards") val attackerRewards: PetCombatRewardsWire,
    @SerialName("defender_rewards") val defenderRewards: PetCombatRewardsWire,
    @SerialName("battle_log") val battleLog: PetBattleLogWire,
    val energy: ProfileEnergyWire? = null
)

@Serializable
data class PetCombatHeadToHeadLastBattleWire(
    @SerialName("created_at") val createdAt: String? = null,
    val won: Boolean = false,
    @SerialName("is_draw") val isDraw: Boolean = false,
    @SerialName("your_role") val yourRole: String? = null,
    val rewards: PetCombatRewardsWire? = null
)

@Serializable
data class PetCombatUserStatsWire(
    @SerialName("max_damage_dealt") val maxDamageDealt: Int = 0,
    @SerialName("max_damage_taken") val maxDamageTaken: Int = 0,
    @SerialName("total_damage_dealt") val totalDamageDealt: Long = 0,
    @SerialName("total_damage_taken") val totalDamageTaken: Long = 0,
    @SerialName("crits_landed") val critsLanded: Int = 0,
    @SerialName("dodges_performed") val dodgesPerformed: Int = 0,
    @SerialName("total_battles") val totalBattles: Int = 0,
    val wins: Int = 0,
    val losses: Int = 0,
    val draws: Int = 0,
    @SerialName("current_win_streak") val currentWinStreak: Int = 0,
    @SerialName("best_win_streak") val bestWinStreak: Int = 0,
    @SerialName("longest_battle_turns") val longestBattleTurns: Int = 0
)

@Serializable
data class PetCombatHeadToHeadSummaryWire(
    val wins: Int = 0,
    val losses: Int = 0,
    val draws: Int = 0,
    @SerialName("total_battles") val totalBattles: Int = 0,
    @SerialName("win_rate") val winRate: Double = 0.0,
    @SerialName("last_battle") val lastBattle: PetCombatHeadToHeadLastBattleWire? = null
) {
    fun winRatePercentText(): String =
        if (totalBattles <= 0) "0%" else "${kotlin.math.round(winRate * 100).toInt()}%"
}

File: Liftr/Pet/PetCombatModels.swift
import Foundation

struct ProfileEnergy: Codable, Equatable {
    let current: Int
    let max: Int
    let lastRefresh: Date?
    let nextRefreshAt: Date?
    let regenMinutes: Int?

    enum CodingKeys: String, CodingKey {
        case current, max
        case lastRefresh = "last_refresh"
        case nextRefreshAt = "next_refresh_at"
        case regenMinutes = "regen_minutes"
    }

    var isFull: Bool { current >= max }
}

enum PetEnergyPricing {
    static let maxCapacity = 15

    static func upgradeCost(maxEnergy: Int) -> Int {
        Int(5000 * pow(2.0, Double(max(maxEnergy, 5) - 5)))
    }
}

struct EnergyUpgradeResult: Codable, Equatable {
    let maxEnergy: Int
    let currentEnergy: Int
    let cost: Int
    let purchaseId: UUID

    enum CodingKeys: String, CodingKey {
        case maxEnergy = "max_energy"
        case currentEnergy = "current_energy"
        case cost
        case purchaseId = "purchase_id"
    }
}

struct PetCombatRewards: Codable, Equatable {
    let xp: Int
    let coins: Int
}

struct PetCombatPetSummary: Codable, Equatable {
    let id: UUID
    let customName: String?
    let petType: String
    let evolutionStage: String
    let currentLevel: Int
    let rarity: String
    let imageUrl: String?
    let name: String?

    enum CodingKeys: String, CodingKey {
        case id
        case customName = "custom_name"
        case petType = "pet_type"
        case evolutionStage = "evolution_stage"
        case currentLevel = "current_level"
        case rarity
        case imageUrl = "image_url"
        case name
    }

    var displayName: String {
        if let name, !name.isEmpty { return name }
        if let customName, !customName.isEmpty { return customName }
        return petType.petDisplayTitle
    }

    var imageURL: URL? {
        if let imageUrl, !imageUrl.isEmpty, let url = URL(string: imageUrl) {
            return url
        }
        return PetImageURLBuilder.imageURL(petType: petType, evolutionStage: evolutionStage)
    }
}

struct PetCombatStatsSummary: Codable, Equatable {
    let health: Int
    let strength: Int
    let defense: Int
    let speed: Int
    let intelligence: Int
    let agility: Int
    let stamina: Int
    let criticalRate: Int
    let resistance: Int
    let exploration: Int
    let happiness: Int

    enum CodingKeys: String, CodingKey {
        case health, strength, defense, speed, intelligence, agility, stamina, resistance, exploration, happiness
        case criticalRate = "critical_rate"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        health = try container.decodeIfPresent(Int.self, forKey: .health) ?? 0
        strength = try container.decodeIfPresent(Int.self, forKey: .strength) ?? 0
        defense = try container.decodeIfPresent(Int.self, forKey: .defense) ?? 0
        speed = try container.decodeIfPresent(Int.self, forKey: .speed) ?? 0
        intelligence = try container.decodeIfPresent(Int.self, forKey: .intelligence) ?? 0
        agility = try container.decodeIfPresent(Int.self, forKey: .agility) ?? 0
        stamina = try container.decodeIfPresent(Int.self, forKey: .stamina) ?? 0
        criticalRate = try container.decodeIfPresent(Int.self, forKey: .criticalRate) ?? 0
        resistance = try container.decodeIfPresent(Int.self, forKey: .resistance) ?? 0
        exploration = try container.decodeIfPresent(Int.self, forKey: .exploration) ?? 0
        happiness = try container.decodeIfPresent(Int.self, forKey: .happiness) ?? 0
    }

    var combatStatPoolTotal: Int {
        health + strength + defense + speed + agility + stamina + resistance + criticalRate + intelligence + exploration
    }

    var comparisonStatPoolTotal: Int {
        Int(floor(Double(health) * 0.25))
            + strength + defense + speed + agility + stamina + resistance + criticalRate + intelligence + exploration
    }
}

enum PetCombatStatBalancing {
    private static let handicapThresholdPct = 125

    static func isUnbalanced(attacker: PetCombatStatsSummary, defender: PetCombatStatsSummary) -> Bool {
        let poolA = attacker.comparisonStatPoolTotal
        let poolB = defender.comparisonStatPoolTotal
        let stronger = max(poolA, poolB)
        let weaker = min(poolA, poolB)
        return stronger * 100 > weaker * handicapThresholdPct
    }
}

struct PetCombatStatBalancingSummary: Codable, Equatable {
    let isUnbalanced: Bool
    let attackerStatPool: Int
    let defenderStatPool: Int
    let attackerPoolBattle: Int?
    let defenderPoolBattle: Int?
    let strongerSide: String?
    let hardcoreBuffMultiplier: Double?
    let hardcoreBonusPercent: Int?

    enum CodingKeys: String, CodingKey {
        case isUnbalanced = "is_unbalanced"
        case attackerStatPool = "attacker_stat_pool"
        case defenderStatPool = "defender_stat_pool"
        case attackerPoolBattle = "attacker_pool_battle"
        case defenderPoolBattle = "defender_pool_battle"
        case strongerSide = "stronger_side"
        case hardcoreBuffMultiplier = "hardcore_buff_multiplier"
        case hardcoreBonusPercent = "hardcore_bonus_percent"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        isUnbalanced = try container.decodeIfPresent(Bool.self, forKey: .isUnbalanced) ?? false
        attackerStatPool = try container.decodeIfPresent(Int.self, forKey: .attackerStatPool) ?? 0
        defenderStatPool = try container.decodeIfPresent(Int.self, forKey: .defenderStatPool) ?? 0
        attackerPoolBattle = try container.decodeIfPresent(Int.self, forKey: .attackerPoolBattle)
        defenderPoolBattle = try container.decodeIfPresent(Int.self, forKey: .defenderPoolBattle)
        strongerSide = try container.decodeIfPresent(String.self, forKey: .strongerSide)
        hardcoreBuffMultiplier = try container.decodeIfPresent(Double.self, forKey: .hardcoreBuffMultiplier)
        hardcoreBonusPercent = try container.decodeIfPresent(Int.self, forKey: .hardcoreBonusPercent)
    }
}

enum PetCombatChallengeMode: String, CaseIterable, Identifiable {
    case balanced
    case hardcore

    var id: String { rawValue }

    var disableNerfChoice: Bool { self == .hardcore }

    var title: String {
        switch self {
        case .balanced: return "Balanced Mode (Safe)"
        case .hardcore: return "Hardcore Mode (No Nerf)"
        }
    }
}

struct PetCombatSide: Codable, Equatable {
    let userId: UUID
    let username: String?
    let pet: PetCombatPetSummary?
    let stats: PetCombatStatsSummary?

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case username, pet, stats
    }
}

struct PetCombatPreview: Codable, Equatable {
    let canChallenge: Bool
    let blockReason: String?
    let cooldownExpiresAt: Date?
    let statBalancing: PetCombatStatBalancingSummary?
    let attacker: PetCombatSide?
    let defender: PetCombatSide?
    let energy: ProfileEnergy?

    enum CodingKeys: String, CodingKey {
        case canChallenge = "can_challenge"
        case blockReason = "block_reason"
        case cooldownExpiresAt = "cooldown_expires_at"
        case statBalancing = "stat_balancing"
        case attacker, defender, energy
    }

    var isStatUnbalanced: Bool {
        if let statBalancing {
            return statBalancing.isUnbalanced
        }
        guard let attackerStats = attacker?.stats, let defenderStats = defender?.stats else {
            return false
        }
        return PetCombatStatBalancing.isUnbalanced(attacker: attackerStats, defender: defenderStats)
    }

    var isAttackerUnderdog: Bool {
        statBalancing?.strongerSide == "defender"
    }

    var hardcoreBonusLabel: String? {
        guard let percent = statBalancing?.hardcoreBonusPercent, percent > 0 else { return nil }
        return "+\(percent)% Coins & XP"
    }

    var effectiveCanChallenge: Bool {
        if canChallenge { return true }
        if blockReason == "cooldown_active",
           let cooldownExpiresAt,
           cooldownExpiresAt <= Date() {
            return true
        }
        return false
    }

    var blockReasonText: String? {
        guard let blockReason else { return nil }
        switch blockReason {
        case "self_challenge": return "You cannot challenge yourself."
        case "attacker_no_pet": return "You need a hatched pet to challenge."
        case "defender_no_pet": return "This user has no pet to challenge."
        case "attacker_egg": return "Your pet must hatch before battling."
        case "defender_egg": return "This user's pet has not hatched yet."
        case "no_energy": return "No arena energy left. You regain 1 energy every 4 hours."
        case "cooldown_active":
            if let cooldownExpiresAt, cooldownExpiresAt > Date() {
                let formatter = RelativeDateTimeFormatter()
                return "You can challenge again \(formatter.localizedString(for: cooldownExpiresAt, relativeTo: Date()))."
            }
            if cooldownExpiresAt == nil {
                return "You recently battled this user. Try again later."
            }
            return nil
        default: return blockReason.replacingOccurrences(of: "_", with: " ").capitalized
        }
    }
}

struct PetBattlePetSnapshot: Codable, Equatable {
    let userId: UUID
    let petInstanceId: UUID
    let name: String
    let petType: String
    let evolutionStage: String
    let level: Int
    let rarity: String
    let imageUrl: String?
    let maxHp: Int?

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case petInstanceId = "pet_instance_id"
        case name
        case petType = "pet_type"
        case evolutionStage = "evolution_stage"
        case level, rarity
        case imageUrl = "image_url"
        case maxHp = "max_hp"
    }

    func resolvedMaxHp(fallback: Int = 1) -> Int {
        if let maxHp, maxHp > 0 { return maxHp }
        return max(fallback, 1)
    }

    var imageURL: URL? {
        if let imageUrl, !imageUrl.isEmpty, let url = URL(string: imageUrl) {
            return url
        }
        return PetImageURLBuilder.imageURL(petType: petType, evolutionStage: evolutionStage)
    }
}

struct PetBattleTurn: Codable, Equatable, Identifiable {
    let turn: Int
    let actor: String
    let action: String
    let damage: Int
    let isCritical: Bool
    let attackerHpAfter: Int
    let defenderHpAfter: Int
    let message: String

    var id: Int { turn }

    enum CodingKeys: String, CodingKey {
        case turn, actor, action, damage, message
        case isCritical = "is_critical"
        case attackerHpAfter = "attacker_hp_after"
        case defenderHpAfter = "defender_hp_after"
    }
}

struct PetBattleResultSummary: Codable, Equatable {
    let winner: String?
    let winnerUserId: UUID?
    let totalTurns: Int
    let isDraw: Bool

    enum CodingKeys: String, CodingKey {
        case winner
        case winnerUserId = "winner_user_id"
        case totalTurns = "total_turns"
        case isDraw = "is_draw"
    }
}

struct PetBattleLog: Codable, Equatable {
    let version: Int
    let attackerPet: PetBattlePetSnapshot
    let defenderPet: PetBattlePetSnapshot
    let turns: [PetBattleTurn]
    let result: PetBattleResultSummary

    enum CodingKeys: String, CodingKey {
        case version, turns, result
        case attackerPet = "attacker_pet"
        case defenderPet = "defender_pet"
    }
}

struct PetCombatResult: Codable, Equatable {
    let combatId: UUID
    let winnerUserId: UUID?
    let attackerRewards: PetCombatRewards
    let defenderRewards: PetCombatRewards
    let battleLog: PetBattleLog
    let energy: ProfileEnergy?

    enum CodingKeys: String, CodingKey {
        case combatId = "combat_id"
        case winnerUserId = "winner_user_id"
        case attackerRewards = "attacker_rewards"
        case defenderRewards = "defender_rewards"
        case battleLog = "battle_log"
        case energy
    }
}

struct PetCombatUserStats: Codable, Equatable {
    let maxDamageDealt: Int
    let maxDamageTaken: Int
    let totalDamageDealt: Int
    let totalDamageTaken: Int
    let critsLanded: Int
    let dodgesPerformed: Int
    let totalBattles: Int
    let wins: Int
    let losses: Int
    let draws: Int
    let currentWinStreak: Int
    let bestWinStreak: Int
    let longestBattleTurns: Int

    enum CodingKeys: String, CodingKey {
        case maxDamageDealt = "max_damage_dealt"
        case maxDamageTaken = "max_damage_taken"
        case totalDamageDealt = "total_damage_dealt"
        case totalDamageTaken = "total_damage_taken"
        case critsLanded = "crits_landed"
        case dodgesPerformed = "dodges_performed"
        case totalBattles = "total_battles"
        case wins, losses, draws
        case currentWinStreak = "current_win_streak"
        case bestWinStreak = "best_win_streak"
        case longestBattleTurns = "longest_battle_turns"
    }
}

struct PetCombatHeadToHeadLastBattle: Codable, Equatable {
    let createdAt: Date?
    let won: Bool
    let isDraw: Bool
    let yourRole: String?
    let rewards: PetCombatRewards?

    enum CodingKeys: String, CodingKey {
        case createdAt = "created_at"
        case won
        case isDraw = "is_draw"
        case yourRole = "your_role"
        case rewards
    }
}

struct PetCombatHeadToHeadSummary: Codable, Equatable {
    let wins: Int
    let losses: Int
    let draws: Int
    let totalBattles: Int
    let winRate: Double
    let lastBattle: PetCombatHeadToHeadLastBattle?

    enum CodingKeys: String, CodingKey {
        case wins, losses, draws
        case totalBattles = "total_battles"
        case winRate = "win_rate"
        case lastBattle = "last_battle"
    }

    var winRatePercentText: String {
        guard totalBattles > 0 else { return "0%" }
        return "\(Int((winRate * 100).rounded()))%"
    }
}
